include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h

```diff
@@ -0,0 +1,661 @@
+    }
+
+
+
+    // Information on the matrix
+
+    template <typename Number, typename Node>
+    inline unsigned int
+    SparseMatrix<Number, Node>::local_size() const
+    {
+      return matrix->getLocalNumRows();
+    }
+
+
+
+    template <typename Number, typename Node>
+    inline std::pair<typename SparseMatrix<Number, Node>::size_type,
+                     typename SparseMatrix<Number, Node>::size_type>
+    SparseMatrix<Number, Node>::local_range() const
+    {
+      size_type begin, end;
+      begin = matrix->getRowMap()->getMinLocalIndex();
+      end   = matrix->getRowMap()->getMaxLocalIndex() + 1;
+
+      return std::make_pair(begin, end);
+    }
+
+
+
+    template <typename Number, typename Node>
+    inline size_t
+    SparseMatrix<Number, Node>::n_nonzero_elements() const
+    {
+      return matrix->getGlobalNumEntries();
+    }
+
+
+
+    template <typename Number, typename Node>
+    MPI_Comm
+    SparseMatrix<Number, Node>::get_mpi_communicator() const
+    {
+      return Utilities::Trilinos::teuchos_comm_to_mpi_comm(matrix->getComm());
+    }
+
+
+
+    // Modifying entries
+
+    template <typename Number, typename Node>
+    SparseMatrix<Number, Node> &
+    SparseMatrix<Number, Node>::operator=(const double d)
+    {
+      Assert(d == 0, ExcScalarAssignmentOnlyForZeroValue());
+
+      if (compressed)
+        {
+          matrix->resumeFill();
+          compressed = false;
+        }
+
+      matrix->setAllToScalar(d);
+
+      return *this;
+    }
+
+
+
+    template <typename Number, typename Node>
+    SparseMatrix<Number, Node> &
+    SparseMatrix<Number, Node>::operator*=(const Number a)
+    {
+      matrix->scale(a);
+      return *this;
+    }
+
+
+
+    template <typename Number, typename Node>
+    SparseMatrix<Number, Node> &
+    SparseMatrix<Number, Node>::operator/=(const Number a)
+    {
+      Assert(a != 0, ExcDivideByZero());
+
+      const Number factor = 1.0 / a;
+      matrix->scale(factor);
+      return *this;
+    }
+
+
+
+    template <typename Number, typename Node>
+    void
+    SparseMatrix<Number, Node>::add(const size_type       row,
+                                    const size_type       n_cols,
+                                    const size_type      *col_indices,
+                                    const TrilinosScalar *values,
+                                    const bool            elide_zero_values,
+                                    const bool /*col_indices_are_sorted*/)
+    {
+      AssertIndexRange(row, this->m());
+
+      // If the matrix is marked as compressed, we need to
+      // call resumeFill() first.
+      if (compressed || matrix->isFillComplete())
+        {
+          matrix->resumeFill();
+          compressed = false;
+        }
+
+      // count zero entries;
+      size_t n_zero_entries = 0;
+      if (elide_zero_values)
+        for (size_t i = 0; i < n_cols; ++i)
+          if (values[i] == 0)
+            ++n_zero_entries;
```

Comment:
You can probably do this more easily readable using
```
  const size_t n_nonzero_entries = (elide_zero_values ?
                               std::count_if (values, values+n_cols, Number(0)) : 0);
```

Reply:
Thank you for that hint!

I slightly adopted that:
```
  const size_t n_zero_entries = (elide_zero_values ?
                               std::count(values, values+n_cols, Number(0)) : 0);
```